src/test/ui/did_you_mean/bad-assoc-pat.stderr

```diff
@@ -0,0 +1,38 @@
+error: missing angle brackets in associated item path
+  --> $DIR/bad-assoc-pat.rs:13:9
+   |
+13 |         [u8]::AssocItem => {}
+   |         ^^^^^^^^^^^^^^^ help: try: `<[u8]>::AssocItem`
```

Comment:
Suggestions are reserved for cases where our level of certainty that they're correct is very high (sometimes going to the length of continuing parsing to verify our theory). Because this error happens at the parser level, we don't know whether `AssocItem` _is_ an associated item or not, so normally we would suggest using a `span_label` instead. Despite this, in this case I'm ok with the use of a `span_suggestion`, but I would reword it to something more descriptive, like "if you meant to use an associated item, surround the type with brackets:", or something shorter (long suggestion messages make the suggestion be presented on it's own).

Reply:
IIRC, @oli-obk in the recent months moved everything containing code snippets to `span_suggestion`, regardless of suggestion reliability.
(FWIW, I'd also estimate this recovery as pretty reliable.)

>"if you meant to use an associated item, surround the type with brackets:"

The primary message already says roughly this, so I avoided repeating it in the label (I also like really short labels in general).

Reply:
I'm working on suggestion levels. Just make everything a suggestion and let me worry about the probablitity of it being correct

Reply:
@oli-obk I've been thinking about making the error more structured, basically having a struct for each possible error the compiler generates, that way we can start playing with having the `--explain` content inline with the actual code being compiled ([similar output to what Elm produces](https://github.com/rust-lang/rfcs/blob/master/text/1644-default-and-expanded-rustc-errors.md#alternatives)). This would require suggestions to be reworked slightly as well. What do you think?

Reply:
Making the errors more explicit is definitely a good idea. Currently analysis code and error reporting is quite mangled. Maybe we can get it separated with this method.

The most important thing is that the diagnostics API stays as usable as it is now. Making it more complex will only unnecessarily hold up regular development

Reply:
@oli-obk I definitely agree. The only thing that worries me is decoupling the place where the error is emitted and where it is "defined" in different files might lead to harder maintenance, but the way I'm thinking about it it should still allow the current ad-hoc `DiagnosticBuilder` usage to keep the current flexibility (and allowing a gradual migration to these structured errors).